src/mds/MDSDaemon.cc

```diff
@@ -0,0 +1,1318 @@
+    suicide();
+  }
+  else if (args[0] == "respawn") {
+    respawn();
+  }
+  else if (args[0] == "cpu_profiler") {
+    ostringstream ss;
+    cpu_profiler_handle_command(args, ss);
+    clog->info() << ss.str();
+  }
+  else if (args[0] == "heap") {
+    if (!ceph_using_tcmalloc())
+      clog->info() << "tcmalloc not enabled, can't use heap profiler commands\n";
+    else {
+      ostringstream ss;
+      vector<std::string> cmdargs;
+      cmdargs.insert(cmdargs.begin(), args.begin()+1, args.end());
+      ceph_heap_profiler_handle_command(cmdargs, ss);
+      clog->info() << ss.str();
+    }
+  } else {
+    if (!(mds_rank && mds_rank->handle_command_legacy(args))) {
+      dout(0) << "unrecognized command! " << args << dendl;
+    }
+  }
+
+  return 0;
+}
+
+/* This function deletes the passed message before returning. */
+
+void MDSDaemon::handle_mds_map(MMDSMap *m)
+{
+  version_t epoch = m->get_epoch();
+  dout(5) << "handle_mds_map epoch " << epoch << " from " << m->get_source() << dendl;
+
+  // is it new?
+  if (epoch <= mdsmap->get_epoch()) {
+    dout(5) << " old map epoch " << epoch << " <= " << mdsmap->get_epoch() 
+	    << ", discarding" << dendl;
+    m->put();
+    return;
+  }
+
+  // keep old map, for a moment
+  MDSMap *oldmap = mdsmap;
+  const MDSMap::DaemonState new_state = mdsmap->get_state_gid(mds_gid_t(monc->get_global_id()));
+  const int incarnation = mdsmap->get_inc_gid(mds_gid_t(monc->get_global_id()));
+  entity_addr_t addr;
+
+  // decode and process
+  mdsmap = new MDSMap;
+  mdsmap->decode(m->get_encoded());
+
+  monc->sub_got("mdsmap", mdsmap->get_epoch());
+
+  // Calculate my effective rank (either my owned rank or my
+  // standby_for_rank if in standby replay)
+  mds_rank_t whoami = mdsmap->get_rank_gid(mds_gid_t(monc->get_global_id()));
+
+  // verify compatset
+  CompatSet mdsmap_compat(get_mdsmap_compat_set_all());
+  dout(10) << "     my compat " << mdsmap_compat << dendl;
+  dout(10) << " mdsmap compat " << mdsmap->compat << dendl;
+  if (!mdsmap_compat.writeable(mdsmap->compat)) {
+    dout(0) << "handle_mds_map mdsmap compatset " << mdsmap->compat
+	    << " not writeable with daemon features " << mdsmap_compat
+	    << ", killing myself" << dendl;
+    suicide();
+    goto out;
+  }
+
+  // If I was put into standby replay, but I am configured for a different standby
+  // type, ignore the map's state and request my standby type (only used
+  // for oneshot replay?)
+  if (new_state == MDSMap::STATE_STANDBY_REPLAY) {
+    if (standby_type != MDSMap::STATE_NULL && standby_type != MDSMap::STATE_STANDBY_REPLAY) {
+      beacon.set_want_state(mdsmap, standby_type);
+      beacon.send();
+      goto out;
+    }
+  }
+
+  if (whoami == MDS_RANK_NONE && (
+      new_state == MDSMap::STATE_STANDBY_REPLAY || new_state == MDSMap::STATE_ONESHOT_REPLAY)) {
+    whoami = mdsmap->get_mds_info_gid(mds_gid_t(monc->get_global_id())).standby_for_rank;
+  }
+
+  // mark down any failed peers
```

Comment:
Why did you move this block down? (It used to be above the standby machinations.)